### PR TITLE
Close #8: Optimized and refactored solver

### LINKED ISSUE #8: Optimize and refactor solver

### PR DESCRIPTION
**Pull request by ognick**

### CODE REVIEW

File: src/main/scala/wop/game/ai/WopSolver.scala [hunk 1]
```diff
@@ -96,43 +104,51 @@ object WopSolver {
...
-          case is: InProgress =>
-            val alphaBetaSolver = AlphaBetaSolver(is.player, ab)
-            @tailrec def processChild(points: List[Point], player: Player, bestEval: EvalPoint): EvalPoint =
-              points match {
-                case (point: Point) :: tl =>
-                  val nextPoint = prevPoint match {
-                    case None => Some(point)
-                    case _ => prevPoint
+  def minMax(state: WopState): Option[Point] = {
+    def loop(state: Either[Foul, WopState], currPoint: Option[Point], depth: Int, ab: Option[(Int, Int)]): EvalPoint =
```
Comment (fomkin) [inline src/main/scala/wop/game/ai/WopSolver.scala:116]: Should it be `@tailrec`?

File: src/main/scala/wop/game/ai/WopSolver.scala [hunk 2]
```diff
@@ -96,43 +104,51 @@ object WopSolver {
...
+                }
+              case _ => returnResult
             }
-          case _ => returnResult
-        }
 
-    case _ => (0, None)
+        case _ => (0, None)
+      }
+    loop(Right(state), None, 0, None)._2
```
Comment (fomkin) [inline src/main/scala/wop/game/ai/WopSolver.scala]:
``` scala
val (_, y) = loop(Right(state), None, 0, None)
y
```

File: src/main/scala/wop/game/ai/WopSolver.scala (hunk 1 above)
Comment (ognick) [reply src/main/scala/wop/game/ai/WopSolver.scala:116]: minMax is not recursive, it is only interface wrapper of loop
and loop can't tailrec optimize

File: src/main/scala/wop/game/ai/WopSolver.scala (hunk 2 above)
Comment (ognick) [reply src/main/scala/wop/game/ai/WopSolver.scala]: fixed

**Comment by fomkin** [comment]
LGTM 👍

**Pull request merged**